File: packages/shared/src/phone.ts
/** Moroccan phone helpers — mirror of jumlaos.shared.phone. */

export class PhoneError extends Error {}

/** Normalize a Moroccan phone number to E.164 (`+212XXXXXXXXX`). */
export function normalizeMa(raw: string): string {
  const digits = raw.replace(/[\s().-]/g, "").replace(/^00/, "+");
  let body: string;
  if (digits.startsWith("+212")) body = digits.slice(4);
  else if (digits.startsWith("212")) body = digits.slice(3);
  else if (digits.startsWith("0")) body = digits.slice(1);
  else throw new PhoneError("unrecognized_phone_format");

  if (!/^\d{9}$/.test(body)) throw new PhoneError("phone_length_invalid");
  if (!/^[567]/.test(body)) throw new PhoneError("not_a_moroccan_mobile");
  return `+212${body}`;
}

export function isMoroccanMobile(e164: string): boolean {
  return /^\+212[67]\d{8}$/.test(e164);
}

File: apps/api/src/jumlaos/shared/phone.py
"""Moroccan phone number normalization.

Accepts the 5 common formats a Moroccan jumala will enter and returns a
canonical E.164 string (`+212XXXXXXXXX`). Everything in the DB is E.164.
"""

from __future__ import annotations

import phonenumbers
from phonenumbers import NumberParseException


class PhoneError(ValueError):
    """Raised for invalid phone numbers."""


def normalize_ma(raw: str) -> str:
    """Normalize any Moroccan-looking number to E.164.

    Examples accepted:
        06 12 34 56 78     ->  +212612345678
        +212 6 12 34 56 78 ->  +212612345678
        00212 6 12345678   ->  +212612345678
        212612345678       ->  +212612345678
        6 12 34 56 78      ->  +212612345678
    """
    if not raw or not isinstance(raw, str):
        raise PhoneError("phone is required")
    cleaned = raw.strip()
    try:
        parsed = phonenumbers.parse(cleaned, "MA")
    except NumberParseException as exc:
        raise PhoneError(f"invalid phone: {raw!r}") from exc
    if not phonenumbers.is_valid_number(parsed):
        raise PhoneError(f"invalid phone: {raw!r}")
    return phonenumbers.format_number(parsed, phonenumbers.PhoneNumberFormat.E164)


def is_moroccan_mobile(e164: str) -> bool:
    """Return True if the E.164 number is a Moroccan mobile (06/07)."""
    if not e164.startswith("+212"):
        return False
    digits = e164.removeprefix("+212")
    return len(digits) == 9 and digits[0] in {"6", "7"}
